python_modules/dagster/dagster/core/definitions/event_metadata/table.py

```diff
@@ -0,0 +1,303 @@
+    in terms of multiple fields (e.g. col_A > col_B) or in terms of rows.
+
+    Args:
+        other (List[str]): Descriptions of arbitrary table-level constraints.
+    """
+
+    def __new__(
+        cls,
+        other: List[str],
+    ):
+        return super(TableConstraints, cls).__new__(
+            cls,
+            other=check.list_param(other, "other", of_type=str),
+        )
+
+
+_DefaultTableConstraints = TableConstraints(other=[])
+
+# ########################
+# ##### TABLE FIELD
+# ########################
+
+
+class _TableFieldSerializer(DefaultNamedTupleSerializer):
+    pass
+
+
+@whitelist_for_serdes(serializer=_TableFieldSerializer)
+class TableField(
+    NamedTuple(
+        "TableField",
+        [
+            ("name", str),
+            ("type", str),
+            ("description", Optional[str]),
+            ("constraints", "TableFieldConstraints"),
+        ],
+    )
+):
+    """Descriptor for a table field. The only property that must be specified
+    by the user is `name`. If no `type` is specified, `string` is assumed. If
+    no `constraints` are specified, the field is assumed to be nullable (i.e. `required = False`)
+    and have no other constraints beyond the data type.
+
+    Args:
+        name (List[str]): Descriptions of arbitrary table-level constraints.
+        type (Optional[str]): The type of the field. Can be an arbitrary
+            string. Defaults to `"string"`.
+        description (Optional[str]): Description of this field. Defaults to `None`.
+        constraints (Optional[TableFieldConstraints]): Field-level constraints.
+            If unspecified, field is nullable with no constraints.
+    """
+
+    def __new__(
+        cls,
+        name: str,
+        type: str = "string",  # pylint: disable=redefined-builtin
+        description: Optional[str] = None,
+        constraints: Optional["TableFieldConstraints"] = None,
+    ):
+        return super(TableField, cls).__new__(
+            cls,
+            name=check.str_param(name, "name"),
+            type=check.str_param(type, "type"),
+            description=check.opt_str_param(description, "description"),
+            constraints=cast(
+                "TableFieldConstraints",
+                check.opt_inst_param(
+                    constraints,
+                    "constraints",
+                    TableFieldConstraints,
+                    default=_DefaultTableFieldConstraints,
+                ),
+            ),
+        )
+
+
+# ########################
+# ##### TABLE FIELD CONSTRAINTS
+# ########################
+
+
+class _TableFieldConstraintsSerializer(DefaultNamedTupleSerializer):
+    pass
+
+
+EnumValue = Union[str, int, float, bool]
+
+
+@whitelist_for_serdes(serializer=_TableFieldConstraintsSerializer)
+class TableFieldConstraints(
+    NamedTuple(
+        "TableFieldConstraints",
+        [
+            ("required", bool),
+            ("unique", bool),
+            ("min_length", Optional[int]),
```

Comment:
Thinking out loud on this: my pre-existing viewpoint has definitely that we should be agnostic to semantics of the constraints, instead of offering particular constraints like "min_length". That said, the semantics of these constraints are fairly unambiguous. I'm still leery though. Here are some of the considerations:
* If we have special-casing of particular constraints, I could see a slippery slope where users beg us to add additional ones or submit PRs to add additional ones.  Even if we decide to tether ourselves to frictionless, the frictionless spec will likely expand over time, and then there will be the question of which version of frictionless each version of Dagster supports.
* Someone who wants to translate their own constraint library to Dagster constraints needs to look at each constraint and understand whether it matches a Dagster constraint, rather than just pass through their string representation.
* People might have specific ways that they want to render common constraints. E.g. someone might want to render a constraint as "minimum: 5", while someone else might want to render the same constraint as ">= 5".

@schrockn might have thoughts.

Reply:
These are all good points. Some responses:

- I too am not particularly pleased with `Frictionless` minimum/maximum since it's not obvious from the term whether it is inclusive or exclusive.
- Since we are renaming to row/column, our rep is moving away from Frictionless anyway.
- Josh's badge constraints were popular in this thread: https://threads.com/34412971011 - Badges done well are going to require some special-casing

In light of the above, why don't we pare back the special cases to `nullable` (formerly `required`; moving away from Frictionless terminology here, `nullable` is clearer IMO) and `unique`. These are standard enough to merit badges. Everything else can be a string (to start)-- maybe we add more special cases with badges going forward.

Reply:
one thought after our meeting: maybe we can follow the `DagsterEventevent_specific_data` pattern here to introduce some kind of field specific data to allow custom/special-casing and make it easier to maintain backcompat.

Reply:
I don't quite follow, could you elaborate or provide an example of what you mean?